src/hotspot/share/compiler/precompiler.cpp

```diff
@@ -56,12 +56,30 @@ class PrecompileIterator : StackObj {
     }
     DirectiveSet* directives = DirectivesStack::getMatchingDirective(methodHandle(_thread, m), nullptr);
     if (directives->DontPrecompileOption) {
-      return false; // excluded
-    } else if (directives->PrecompileRecordedOption > 0) {
+      return false;
+    }
+    if (directives->PrecompileRecordedOption > 0) {
       return true;
     }
-    int cid = compile_id(m, _search_level);
-    return (cid < INT_MAX);
+    int high_top_level = highest_top_level(m);
+    switch (_comp_level) {
+      case CompLevel_simple:
+      case CompLevel_full_optimization:
+        // For final C1/C2 compilations, we only compile when there was relevant compilation during training.
+        return _comp_level == high_top_level;
+      case CompLevel_limited_profile:
+        // For profiled C1 compilations, generate limited profile when there was limited/full
+        // profiled compilation in training.
+        return CompLevel_limited_profile <= high_top_level && high_top_level <= CompLevel_full_profile;
+      case CompLevel_full_profile:
+        // We do not include C1 full profiled methods at this time.
+        // TODO: See if it is profitable to do so. This requires MDO support in AOTCache.
```

Comment:
We already support reference to MDO from C1 compiled code because we cache tier2 which do profiling.
But we may miss few places in tier3 code.

Reply:
I thought we only record MCS, and reinstantiating MDO. But my recollection is vague. I dropped this to avoid confusion.